### PR TITLE
Event V2

### PR DESCRIPTION
**Pull request by macpie**
```
{
    "id": "UUID-V4",
    "category": "uplink | downlink | join_request | join_accept | misc",
    "sub_category": "undefined | uplink_confirmed | uplink_unconfirmed | uplink_integration_req | uplink_integration_res | uplink_dropped | downlink_confirmed | downlink_unconfirmed | downlink_dropped | downlink_queued | downlink_ack | misc_integration_error",
    "description": "description",
    "reported_at": 123, // in MS
    "device_id": "device_uuid",
    "data": {
        // Start Sub-Category=uplink_confirmed/uplink_unconfirmed/uplink_dropped/downlink_confirmed/downlink_unconfirmed/downlink_ack
        "fcnt": 2,
        "payload_size": 12,
        "payload": "base64 payload", 
        "port": 1,
        "devaddr": "devaddr",
        "hotspot": {
            "id": "hotspot_id",
            "name": "hotspot name",
            "rssi": -30,
            "snr": 0.2,
            "spreading": "SF9BW125",
            "frequency": 923.3,
            "channel": 12,
            "lat": 37.00001962582851,
            "long": -120.9000053210367
        },
        // This is only for category uplink_confirmed/uplink_unconfirmed
        "dc": {
            "balance": 3000,
            "nonce": 2,
            "used": 1
        },
        // This is only for category downlink stuff
        "integration": {
            "id": "integration_id",
            "name": "integration_name",
            "status": "success | error"
        },
        // END Sub-Category=uplink_confirmed/uplink_unconfirmed/uplink_dropped/downlink_confirmed/downlink_unconfirmed/downlink_ack

        // Start Sub-Category=uplink_integration_req/uplink_integration_res/misc_integration_error/downlink_dropped/downlink_queued
        "integration": {
            "id": "integration_id",
            "name": "integration_name",
            "status": "success | error",
        },
        // END Sub-Category=uplink_integration_req/uplink_integration_res/misc_integration_error/downlink_dropped/downlink_queued
        // Start Sub-Category=uplink_integration_req
        "req": {},
         // End Sub-Category=uplink_integration_req
         // Start Sub-Category=uplink_integration_res
        "res": {},
         // End Sub-Category=uplink_integration_res
        
    }
}
```

### CODE REVIEW

File: src/channels/router_http_channel.erl [hunk 1]
```diff
@@ -174,64 +182,92 @@ is_non_local_address(Host) ->
...
+    HeliumError :: {error, atom()},
+    HackneyResponse :: {ok, any()}.
+make_response_report({error, Reason}, Channel) ->
+    #{
         id => router_channel:id(Channel),
         name => router_channel:name(Channel),
-        reported_at => erlang:system_time(seconds)
+        response => #{},
+        status => error,
+        description => list_to_binary(io_lib:format("Helium Error: ~p", [Reason]))
```
Comment (michaeldjeffrey) [inline src/channels/router_http_channel.erl:223]: Remove specifying error source (Helium, or Hackney)

File: src/router_utils.erl [hunk 2]
```diff
@@ -12,6 +26,252 @@
...
+        status => error,
+        description => Description,
+        reported_at => erlang:system_time(millisecond),
+        %%
+        channel_id => maps:get(id, ChannelInfo),
+        channel_name => maps:get(name, ChannelInfo),
+        channel_status => maps:get(status, ChannelInfo)
+    },
+    ok = router_console_api:event(Device, Map).
+
```
Comment (macpie) [inline src/router_utils.erl:246]: specs

File: src/apis/router_console_api.erl [hunk 3]
```diff
@@ -176,6 +171,105 @@ get_channels(Device, DeviceWorkerPid) ->
...
+                    lager:warning("got non 200 resp ~p", [_Other]),
+                    End = erlang:system_time(millisecond),
+                    ok = router_metrics:console_api_observe(report_status, error, End - Start)
+            end
+        end
+    ),
+    ok.
+
 -spec report_status(Device :: router_device:device(), Map :: map()) -> ok.
 report_status(Device, Map) ->
```
Comment (macpie) [inline src/apis/router_console_api.erl:274]: This will need to be remove once we are done

File: src/channels/router_console_channel.erl [hunk 4]
```diff
@@ -36,24 +33,14 @@ init({[Channel, Device], _}) ->
...
-                status => <<"success">>,
-                description => <<"console debug">>,
-                id => router_channel:id(Channel),
-                name => router_channel:name(Channel),
-                reported_at => erlang:system_time(seconds),
-                debug => #{req => #{body => router_channel:encode_data(Channel, Data)}}
-            },
-            router_device_channels_worker:report_status(Pid, Ref, Report)
-    end,
+handle_event({data, UUIDRef, Data}, #state{channel = Channel} = State) ->
```
Comment (macpie) [inline src/channels/router_console_channel.erl:36]: Some much cleaner without debug I love it

File: src/channels/router_http_channel.erl [hunk 5]
```diff
@@ -174,64 +182,101 @@ is_non_local_address(Host) ->
...
+            });
+        {ok, StatusCode, ResponseHeaders, ResponseBody} when StatusCode >= 200, StatusCode =< 300 ->
+            maps:merge(Result0, #{
+                response => #{
+                    code => StatusCode,
+                    headers => ResponseHeaders,
+                    body => ResponseBody
+                },
+                status => success,
+                description => ResponseBody
```
Comment (macpie) [inline src/channels/router_http_channel.erl:259]: I am not sure about this, what if body is json string do we wanna show that as description?

File: src/device/router_device_channels_worker.erl [hunk 6]
```diff
@@ -184,141 +191,124 @@ init(Args) ->
...
-                                    false ->
-                                        DataCache0
-                                end
-                        end
-                end
+    Action =
+        case maps:get(UUID, DataCache0, undefined) of
+            undefined ->
+                new_uuid;
+            #{PubKeyBin := CachedPacket} = CachedData ->
```
Comment (macpie) [inline src/device/router_device_channels_worker.erl:218]: why are we still doing this here?

File: test/router_channel_console_SUITE.erl [hunk 7]
```diff
@@ -75,27 +75,25 @@ console_test(Config) ->
...
     Stream ! {send, test_utils:join_packet(PubKeyBin, AppKey, DevNonce)},
-    timer:sleep(router_device_utils:join_timeout()),
+    timer:sleep(router_utils:join_timeout()),
 
     %% Waiting for report device status on that join request
-    test_utils:wait_for_console_event(<<"join_req">>, #{
-        <<"category">> => <<"join_req">>,
-        <<"description">> => '_',
+    test_utils:wait_for_console_event(<<"join_request">>, #{
+        <<"id">> => fun erlang:is_binary/1,
```
Comment (macpie) [inline test/router_channel_console_SUITE.erl:82]: Could we somehow here trace that UUID to make sure that it is the same for all related events?

File: test/router_channel_http_SUITE.erl [hunk 8]
```diff
@@ -69,27 +69,25 @@ http_test(Config) ->
...
     Stream ! {send, test_utils:join_packet(PubKeyBin, AppKey, DevNonce)},
-    timer:sleep(router_device_utils:join_timeout()),
+    timer:sleep(router_utils:join_timeout()),
 
     %% Waiting for report device status on that join request
-    test_utils:wait_for_console_event(<<"join_req">>, #{
-        <<"category">> => <<"join_req">>,
-        <<"description">> => '_',
+    test_utils:wait_for_console_event(<<"join_request">>, #{
+        <<"id">> => fun erlang:is_binary/1,
```
Comment (macpie) [inline test/router_channel_http_SUITE.erl:76]: Same as console suite for UUID

File: test/router_device_channels_worker_SUITE.erl [hunk 9]
```diff
@@ -562,20 +600,12 @@ late_packet_test(Config) ->
...
-                <<"status">> => <<"success">>,
-                <<"description">> => '_'
-            }
-        ]
+        }
     }),
 
     %% Make sure we did not get a duplicate with that late message
     receive
+        %% REVIEW: Wrong shape, what messages do we actually want to break the test?
```
Comment (macpie) [inline test/router_device_channels_worker_SUITE.erl:608]: ?

File: test/utils/test_utils.erl [hunk 10]
```diff
@@ -256,6 +257,28 @@ wait_for_console_event(Category, Expected) ->
             ct:fail("wait_for_console_event ~p failed", [Category])
     end.
 
+wait_for_console_event_sub(SubCategory, Expected) ->
```
Comment (macpie) [inline test/utils/test_utils.erl:260]: Maybe returning UUID here for later matching (see comments above)

File: src/channels/router_http_channel.erl (hunk 5 above)
Comment (michaeldjeffrey) [reply src/channels/router_http_channel.erl:259]: Let's got with generic descriptions for now.

File: src/device/router_device_channels_worker.erl (hunk 6 above)
Comment (michaeldjeffrey) [reply src/device/router_device_channels_worker.erl]: Let's always take packets here. Don't care about stronger or weaker rssi

File: test/router_device_channels_worker_SUITE.erl (hunk 9 above)
Comment (michaeldjeffrey) [reply test/router_device_channels_worker_SUITE.erl:567]: get rid of this `receive` block. no longer makes sense in this test.

**Review by vicmgs — APPROVED** [review]

**Pull request merged**